array/array_test.go

```diff
@@ -0,0 +1,115 @@
+package array
+
+import (
+	"testing"
+)
+
+func TestArray(t *testing.T) {
+	t.Run("Create new array with default capacity", func(t *testing.T) {
+		arr := NewArray()
+		if capacity := arr.Capacity(); capacity != defaultCapacity {
+			t.Errorf("New array capacity should equal %d, found %d", defaultCapacity, capacity)
+		}
+	})
+
+	t.Run("Create new array with custom capacity", func(t *testing.T) {
+		customCapacity := 10
+		arr := NewArray(customCapacity)
+		if capacity := arr.Capacity(); capacity != customCapacity {
+			t.Errorf("New array capacity should equal %d, found %d", customCapacity, capacity)
+		}
+	})
+
+	arr := NewArray()
+
+	t.Run("Check size for empty array", func(t *testing.T) {
+		if arr.Size() != 0 || arr.IsEmpty() != true {
+			t.Error("Empty array should have 0 size")
+		}
+	})
+
+	t.Run("Get value from empty array, get indexOutOfRange", func(t *testing.T) {
+		if _, err := arr.ValueAt(0); err == nil {
+			t.Error("Calling ValueAt on empty array should return indexOutOfRange")
+		}
+	})
+
+	t.Run("Insert value to out-of-range index, get indexOutOfRange", func(t *testing.T) {
+		if err := arr.Insert(1, 1); err == nil {
+			t.Error("Insert to index 1 on empty array should return indexOutOfRange")
+		}
+	})
+
+	t.Run("Remove value at out-of-range index, get indexOutOfRange", func(t *testing.T) {
+		if err := arr.RemoveAt(1); err == nil {
+			t.Error("Remove value at index 1 on empty array should return indexOutOfRange")
+		}
+	})
+
+	t.Run("Pop value from empty array, get EmptyValue", func(t *testing.T) {
+		if value := arr.Pop(); value != EmptyValue {
+			t.Error("Value popped from empty array should be EmptyValue")
+		}
+	})
+
+	t.Run("Append values to array, get correct appended values", func(t *testing.T) {
+		for i := 1; i < 4; i++ {
+			arr.Append(i)
+
+			if value, _ := arr.ValueAt(i - 1); value != i {
+				t.Errorf("Wrong value at index %d, expected %d, found %d", i-1, i, value)
+			}
+		}
+	})
+
+	t.Run("Insert value at specified index, get correct inserted value", func(t *testing.T) {
+		arr.Insert(4, 0)
+
+		for i := 0; i < arr.Size(); i++ {
+			value, _ := arr.ValueAt(i)
+			if i == 0 {
+				if value != 4 {
+					t.Errorf("Wrong value at index 0, expected 4, found %d", value)
+				}
+			} else {
+				if value != i {
+					t.Errorf("Wrong value at index %d, expected %d, found %d", i, i, value)
+				}
+			}
+		}
+	})
+
+	t.Run("Append more values so array must grow in capacity", func(t *testing.T) {
+		t.Logf("Capacity before: %d", arr.Capacity())
+
+		for i := 4; i < 20; i++ {
+			arr.Append(i)
+		}
+
+		t.Logf("Capacity after: %d", arr.Capacity())
```

Comment:
We will not log anything in test cases cause noone will read it. Instead of logging capacity, it's better to check capacity is equal a value. If it's not equal expected value, test case fails.